08-CMSIS/core/src/threads.c

```diff
@@ -0,0 +1,131 @@
+#include <stdint.h>
+#include "threads.h"
+#include "os.h"
+#include "malloc.h"
+#include "reg.h"
+
+#define THREAD_PSP	0xFFFFFFFD
+
+/* Thread Control Block */
+typedef struct {
+	void *stack;
+	void *orig_stack;
+	uint8_t in_use;
+} tcb_t;
+
+static tcb_t tasks[MAX_TASKS];
+static int lastTask;
+static int first = 1;
+
+/* FIXME: Without naked attribute, GCC will corrupt r7 which is used for stack
+ * pointer. If so, after restoring the tasks' context, we will get wrong stack
+ * pointer.
+ */
+void __attribute__((naked)) pendsv_handler()
+{
+	/* Save the old task's context */
+	asm volatile("mrs   r0, psp\n"
+	             "stmdb r0!, {r4-r11, lr}\n");
+	/* To save the last task's sp from r0 into its tcb*/
+	asm volatile("mov   %0, r0\n" : "=r"(tasks[lastTask].stack));
+
+	/* Find a new task to run */
+	while (1) {
+		lastTask++;
+		if (lastTask == MAX_TASKS)
+			lastTask = 0;
+		if (tasks[lastTask].in_use) {
+			/* Move the task's stack pointer address into r0 */
+			asm volatile("mov r0, %0\n" : : "r"(tasks[lastTask].stack));
+			/* Restore the new task's context and jump to the task */
+			asm volatile("ldmia r0!, {r4-r11, lr}\n"
+			             "msr psp, r0\n"
+			             "bx lr\n");
+		}
+	}
+}
+
+
+void thread_start()
+{
+	lastTask = 0;
+
+	/* Save kernel context */
+	asm volatile("mrs ip, psr\n"
+	             "push {r4-r11, ip, lr}\n");
+
+	/* To bridge the variable in C and the register in ASM,
+	 * move the task's stack pointer address into r0.
+	 * http://www.ethernut.de/en/documents/arm-inline-asm.html
+	 */
+	asm volatile("mov r0, %0\n" : : "r"(tasks[lastTask].stack));
+	/* Load user task's context and jump to the task */
+	asm volatile("msr psp, r0\n"
```

Comment:
`core_cmFunc.h` stands for core peripheral function, and we can invoke the inline functions without writing inline assembly.